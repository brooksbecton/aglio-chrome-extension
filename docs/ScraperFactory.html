<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Class: ScraperFactory</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Class: ScraperFactory</h1>

    




<section>

<header>
    
        <h2><span class="attribs"><span class="type-signature"></span></span>ScraperFactory<span class="signature">()</span><span class="type-signature"></span></h2>
        
            <div class="class-description">Creates an appropriate scraper for a given url</div>
        
    
</header>

<article>
    <div class="container-overview">
    
        

    
    <h2>Constructor</h2>
    

    
    <h4 class="name" id="ScraperFactory"><span class="type-signature"></span>new ScraperFactory<span class="signature">()</span><span class="type-signature"></span></h4>
    

    















<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="lib_scrapers_ScraperFactory_index.js.html">lib/scrapers/ScraperFactory/index.js</a>, <a href="lib_scrapers_ScraperFactory_index.js.html#line6">line 6</a>
    </li></ul></dd>
    

    

    

    
</dl>


















    
    </div>

    

    

    

    

    

    

    

    
        <h3 class="subsection-title">Methods</h3>

        
            

    

    
    <h4 class="name" id="contains"><span class="type-signature"></span>contains<span class="signature">(testStr, searchStr)</span><span class="type-signature"></span></h4>
    

    



<div class="description">
    Searches for sub-string and returns true if testStr contains it
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>testStr</code></td>
            

            <td class="type">
            
                
<span class="param-type">String</span>


            
            </td>

            

            

            <td class="description last">the string to look through</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>searchStr</code></td>
            

            <td class="type">
            
                
<span class="param-type">String</span>


            
            </td>

            

            

            <td class="description last">the query to be searched for</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="lib_scrapers_ScraperFactory_index.js.html">lib/scrapers/ScraperFactory/index.js</a>, <a href="lib_scrapers_ScraperFactory_index.js.html#line12">line 12</a>
    </li></ul></dd>
    

    

    

    
</dl>


















        
            

    

    
    <h4 class="name" id="create"><span class="type-signature">(async) </span>create<span class="signature">()</span><span class="type-signature"> &rarr; {function}</span></h4>
    

    



<div class="description">
    Creates an instance of a scraper based on the url supplied
</div>













<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="lib_scrapers_ScraperFactory_index.js.html">lib/scrapers/ScraperFactory/index.js</a>, <a href="lib_scrapers_ScraperFactory_index.js.html#line22">line 22</a>
    </li></ul></dd>
    

    

    

    
</dl>











<h5>Throws:</h5>

        

    <div class="param-desc">
    
        Will throw an error if it doesn't have a scraper for a url
    
    </div>


    


<h5>Returns:</h5>

        
<div class="param-desc">
    A Child instance of BaseScraper
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">function</span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id="getCurrentTab"><span class="type-signature"></span>getCurrentTab<span class="signature">()</span><span class="type-signature"> &rarr; {Promise}</span></h4>
    

    



<div class="description">
    Gets the user's current page <a href="https://developer.chrome.com/extensions/tabs#type-Tab%20"> tab </a>Using <a href="https://developer.chrome.com/extensions/tabs#method-getCurrent%20"> getCurrent</a>
</div>













<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="lib_scrapers_ScraperFactory_index.js.html">lib/scrapers/ScraperFactory/index.js</a>, <a href="lib_scrapers_ScraperFactory_index.js.html#line38">line 38</a>
    </li></ul></dd>
    

    

    

    
</dl>













<h5>Returns:</h5>

        
<div class="param-desc">
    user's current tab
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">Promise</span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id="getHostname"><span class="type-signature"></span>getHostname<span class="signature">(qUrl)</span><span class="type-signature"> &rarr; {String}</span></h4>
    

    



<div class="description">
    Parses and returns hostname from a given URL
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>qUrl</code></td>
            

            <td class="type">
            
                
<span class="param-type">String</span>


            
            </td>

            

            

            <td class="description last"></td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="lib_scrapers_ScraperFactory_index.js.html">lib/scrapers/ScraperFactory/index.js</a>, <a href="lib_scrapers_ScraperFactory_index.js.html#line56">line 56</a>
    </li></ul></dd>
    

    

    

    
</dl>













<h5>Returns:</h5>

        
<div class="param-desc">
    hostname of the given URL
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">String</span>


    </dd>
</dl>

    





        
    

    

    
</article>

</section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Classes</h3><ul><li><a href="AllRecipesScraper.html">AllRecipesScraper</a></li><li><a href="App.html">App</a></li><li><a href="BaseScraper.html">BaseScraper</a></li><li><a href="Recipe.html">Recipe</a></li><li><a href="RecipeDetails.html">RecipeDetails</a></li><li><a href="SaveRecipeButton.html">SaveRecipeButton</a></li><li><a href="ScraperFactory.html">ScraperFactory</a></li></ul><h3><a href="global.html">Global</a></h3>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.5.5</a> on Tue May 29 2018 10:40:10 GMT-0500 (Central Daylight Time)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>