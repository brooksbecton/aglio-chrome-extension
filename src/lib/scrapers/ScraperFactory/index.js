/*global chrome*/

import AllRecipesScraper from "./../AllRecipesScraper";

/** Creates an appropriate scraper for a given url */
class ScraperFactory {
  /**
   * Searches for sub-string and returns true if testStr contains it
   * @param {String} testStr the string to look through
   * @param {String} searchStr the query to be searched for
   */
  contains(testStr, searchStr) {
    return testStr.indexOf(searchStr) > -1;
  }

  /**
   * Creates an instance of a scraper based on the url supplied
   * @async
   * @throws Will throw an error if it doesn't have a scraper for a url
   * @returns {Function} A Child
   */
  async create() {
    const tab = await this.getCurrentTab();
    const hostname = this.getHostname(tab.url);

    if (this.contains(hostname, "allrecipes")) {
      return new AllRecipesScraper(tab.url);
    } else {
      throw new Error(`No Scraper for ${hostname}`);
    }
  }
  /**
   * Gets the user's current page {@link https://developer.chrome.com/extensions/tabs#type-Tab | tab }
   * Using {@link https://developer.chrome.com/extensions/tabs#method-getCurrent | getCurrent}
   *
   * @returns {Promise} user's current tab
   */
  getCurrentTab() {
    return new Promise((resolve, reject) => {
      chrome.tabs.query({ active: true, currentWindow: true }, tabs => {
        if (tabs.length > 0) {
          resolve(tabs[0]);
        } else {
          reject(new Error("No Tab Found"));
        }
      });
    });
  }

  /**
   * Parses and returns hostname from a given URL
   *
   * @param {String} qUrl
   * @returns {String} hostname of the given URL
   */
  getHostname(qUrl) {
    let hostname;
    try {
      const url = new URL(String(qUrl));
      hostname = url.hostname;
    } catch (err) {
      throw new Error(`${qUrl} is not a valid URL - ${err}`);
    }
    return hostname;
  }
}

export default ScraperFactory;
