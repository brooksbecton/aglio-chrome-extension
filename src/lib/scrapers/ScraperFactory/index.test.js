import ScraperFactory from "./index";
import AllRecipeScraper from "./../AllRecipesScraper";
describe("ScraperFactory", () => {
  it("gets hostname from url", () => {
    const testScraper = new ScraperFactory();

    expect(
      testScraper.getHostname("https://github.com/lifaon74/url-polyfill")
    ).toBe("github.com");
  });

  it("gets returns an instance of 'AllRecipeScraper' for an allrecipe URL", async () => {
    const testScraper = new ScraperFactory();
    testScraper.getCurrentTab = jest.fn(() => {
      return {
        url:
          "https://www.allrecipes.com/recipe/220854/chef-johns-italian-meatballs/"
      };
    });
    expect.assertions(1);

    expect(await testScraper.create()).toBeInstanceOf(AllRecipeScraper);
  });

  it("throws an error if it does not have a scraper for a given url", async () => {
    const testScraper = new ScraperFactory();

    testScraper.getCurrentTab = jest.fn(() => {
      return new Promise(resolve => {
        resolve({
          url: "https://facebook.github.io/jest/docs/en/asynchronous.html"
        });
      });
    });
    try {
      await testScraper.create();
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });
});
