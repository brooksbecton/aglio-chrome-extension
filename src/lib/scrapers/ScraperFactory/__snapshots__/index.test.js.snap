// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScraperFactory throws an error if it does not have a scraper for a given url 1`] = `[Error: No Scraper for facebook.github.io]`;
